cookbook/case_studies/feature_engineering/sqlite_datacleaning/workflow.py

```diff
@@ -0,0 +1,77 @@
+import typing
+
+import pandas as pd
+from flytekit import Workflow, kwtypes, reference_task
+from flytekit.extras.sqlite3.task import SQLite3Config, SQLite3Task
+from flytekit.types.file import FlyteFile
+from flytekit.types.schema import FlyteSchema
+
+
+@reference_task(
+    project="flytesnacks",
+    domain="development",
+    name="sqlite_datacleaning.tasks.mean_median_imputer",
+    version="aedbd6fe44051c171fd966c280c5c3036f658831",
+)
+def mean_median_imputer(
+    dataframe: pd.DataFrame,
+    imputation_method: str,
+) -> pd.DataFrame:
+    ...
+
+
+@reference_task(
+    project="flytesnacks",
+    domain="development",
+    name="sqlite_datacleaning.tasks.univariate_selection",
+    version="aedbd6fe44051c171fd966c280c5c3036f658831",
+)
+def univariate_selection(
+    dataframe: pd.DataFrame,
+    split_mask: int,
+    num_features: int,
+) -> pd.DataFrame:
+    ...
+
+
+wb = Workflow(name="sqlite_datacleaning.workflow.fe_wf")
+wb.add_workflow_input("sqlite_archive", FlyteFile[typing.TypeVar("zip")])
+wb.add_workflow_input("imputation_method", str)
+wb.add_workflow_input("query_template", str)
+
+
+sql_task = SQLite3Task(
+    name="sqlite3.horse_colic",
+    query_template=wb.inputs["query_template"],
```

Comment:
Including workflow inputs in SQLite3Task is resulting in an error during serialization:
`TypeError: Object of type Promise is not JSON serializable` (I believe it works in a typical task)
How do I resolve this?

Reply:
this isn't quite the same thing right as a normal python task right?  You can't pass `wb.inputs[...]` to `SQLite3Task` because the invocation of `SQLite3Task` doesn't call the task, it creates one.  Basically `SQLite3Task(...)` creates a task much like how
```
@task
def t(): ...
```
creates a task, which then has to be called in a workflow somewhere.  The current code is kind of akin to providing a default input to a task (which we don't support yet but may in the future), but one that varies (which doesn't make sense).

The `SQLite3Task` task type must have the query template specified at task creation time, much like how at regular-task creation time, you must specify the inputs/outputs (the function signature).

There is a hack here, which is that you can make the entire template an input:
```
hack_sql_task = SQLite3Task(name="hacky.sample.sql.task", query_template="{{ .inputs.query }}", inputs=kwtypes(query=str), ...)
```
At execution time, the entire sql will be just what the user provided as the workflow input, but you can see how this is kind of a stretch.  Note though that the template itself is a constant string.  I think we wrote these tasks with most of the query filled out in mind.  The template would hardcode say the table and the columns returned and the inputs would only be used to control the limit or the where clause or something
```
query_template = """
select col_a
from table_name
where ds >= "{{ .inputs.start_date }}" and ds < "{{ .inputs.stop_date }}"
limit {{ .inputs.limit }}
"""
or something like that.

I think for purposes of this example, we should have as much of the query set in stone as possible.

Reply:
Thanks, @wild-endeavor! However say, I use `limit` (not the whole query template) as an input. In such a case, `node_t1 = wb.add_entity(sql_task(limit=100))` isn't a possibility in an imperative workflow as `()` runs the task, correct? This also means that in an imperative workflow, SQLite3Task shouldn't accept any user inputs.

Reply:
Yeah calling `()` will run the task. To add a task (or launch plan or subwf) with the `add_entity` call, you need to do
```
node_t1 = wb.add_entity(sql_task, limit=100)
```
that will store the scalar value as a binding.  See the test file for more examples and it even works for more [complicated bindings](https://github.com/flyteorg/flytekit/blob/aec64613684eb26ffd4be99af41353342083231e/tests/flytekit/unit/core/test_imperative.py#L94).

Maybe some of this discussion should be included in the documentation as well?

Reply:
Sure thing. We can do that.